### PR TITLE
Refactor/topic dash controller

### LINKED ISSUE #23: TopicDashReducer Initialization Action Suggestion
Heya Janet! This is an extension of a comment I made on your PR #22 but I think an issue is a better platform to discuss. 

I hope it's okay, but I'd like to offer a suggestion about managing your topicDashReducer state. I think copying this over to your reducer would make your life a lot easier. I've added an example of what your state would now look like at the bottom. 

```js
case RESOURCE_SUCCESS:
  state.loading = false;
  state.error = null;

  // One thing that will make your life 1000000x easier is adding your resources
  // to an object using this function:
  function mapResourcesToObject(resources) {
    return resources.reduce((obj, resource) => {
      obj[resource.id] = resource;
      return obj;
    }, {})
  }

  state.resources = mapResourcesToObject(action.payload.resources);
  // remove redundant resources from payload
  delete action.payload.resources;

  state.resourceOrder = action.payload.resourceOrder;
  // remove redundant resourceOrder from payload
  delete action.payload.resourceOrder;

  state.topic = action.payload;
```

Here's what your state would look like if you adopted this idea:

```js
// EXAMPLE OF YOUR NEW STATE WITH THIS REDUCER

state.topicDashReducer = {
  topic: {
    "id": 4000,
    "title": "Node",
    "parent": {
        "id": 3001,
        "title": "Javascript"
    },
    "createdAt": "2018-12-13T23:39:04.638Z",
    "updatedAt": "2018-12-13T23:39:04.638Z",
  },
  resourceOrder: [1, 2, 3],
  resources: {
  "1": {
      "id": 1,
      "parent": {
          "id": 4000,
          "title": "Node"
      },
      "title": "Resource #1",
      "type": "youtube",
      "uri": "4DfzRRkiPX4",
      "completed": false,
      "lastOpened": "2018-12-12T22:28:18.573Z"
  },
  "2": {
      "id": 2,
      "parent": {
          "id": 4000,
          "title": "Node"
      },
      "title": "Resource #2",
      "type": "youtube",
      "uri": "B7hVxCmfPtM",
      "completed": false,
      "lastOpened": "2018-12-13T14:38:11.895Z"
  },
  "3": {
      "id": 3,
      "parent": {
          "id": 4001,
          "title": "ES6"
      },
      "title": "Resource #3",
      "type": "youtube",
      "uri": "B7hVxCmfPtM",
      "completed": false,
      "lastOpened": "2018-12-13T16:01:29.105Z"
  },
  loading: false,
  error: null
}
```

Notice how you can access your resources in an object now by using their IDs? This is 10x easier for your use case than trying to find it in an array for each and every item. The order is given by the IDs in the resourceOrder array, so you can quite easily load things in order by iterating through the resourceOrder array and look up the resources by ID.

FYI this is an extremely common thing to do. It's called data normalization and it makes your life a lot easier. [Here's a good article on it](https://redux.js.org/recipes/structuring-reducers/normalizing-state-shape#designing-a-normalized-state).

_Originally posted by @morgannewman in https://github.com/morgannewman/learn-note/pull/22_

### PR DESCRIPTION
**Pull request by jileon**
Fixes #23 

This refactors topicDash actions, reducer and components so that data is normalized and can be referenced as objects. It fixes the actions to take resourceOrder into consideration as well as being able to update or delete resources based on the new state structure. Right now it is functional minus the youtube URLs that need some front end handling to that they look like URLs and not codes. I also noticed that the back end no long has 'URI' as an update-able field, are we not doing that anymore? Right now I have it so the resource component is able to edit a URI, but I can easily disable it.

### CODE REVIEW

File: src/app/TopicDasboard/ActiveResourceContainer.js [hunk 1]
```diff
@@ -4,29 +4,38 @@ import AddResourceForm from './AddResourceForm';
 import './ActiveResourceContainer.scss';
 
 export class ActiveResourceContainer extends React.Component {
-	//TODO: Drag and drop functionality
+  //TODO: Drag and drop functionality
+  state = {
+    resourceOrder: this.props.resourceOrder
+  };
+  //FIXME: resources[rescID] doesn't always sync up on delete
```
Comment (morgannewman) [inline src/app/TopicDasboard/ActiveResourceContainer.js]: Is this still an issue?

File: src/controller/actions/topicDashboard.js [hunk 2]
```diff
@@ -20,31 +20,35 @@ export const resourceError = error => ({
...
+export const updateResource = resource => ({
   type: UPDATE_RESOURCE,
-  resource,
-  id
+  payload: resource
 });
 
-export const DELETE_RESOURCE = 'DELETE_RESOURCES';
+export const UPDATE_RESC_ORDER = 'UPDATE_RESC_ORDER';
+export const updateRescourceOrder = rescOrder => ({
```
Comment (morgannewman) [inline src/controller/actions/topicDashboard.js]: Please fix this typo here and all places it's imported.

```js
-  updateRescourceOrder
+  updateResourceOrder
```

File: src/controller/actions/topicDashboard.js [hunk 3]
```diff
@@ -20,31 +20,35 @@ export const resourceError = error => ({
...
+ *Gets topic information for a specified topicID includeing resources, resource order, topic title, and ID
  *First dispatches a loading function to change state to loading
- *Second: Sends GET request to the server using api.resources
+ *Second: Sends GET request to the server using api.topic
  * Third: Dispatches resource_success action to add new data from server to current state
  * If there is an error, it dispatches an error obj to state
  * *@param {{id:integer}}} id
  */
-export const getResources = id => dispatch => {
+export const initializeTopicDasbhoard = id => dispatch => {
```
Comment (morgannewman) [inline src/controller/actions/topicDashboard.js]: Please fix this typo here and all places it's referenced.

```js
-  initializeTopicDasbhoard
+  initializeTopicDashboard
```
```

File: src/app/TopicDasboard/ActiveResourceContainer.js (hunk 1 above)
Comment (jileon) [reply src/app/TopicDasboard/ActiveResourceContainer.js]: NOPE, that's a stale comment

File: src/controller/actions/topicDashboard.js (hunk 2 above)
Comment (jileon) [reply src/controller/actions/topicDashboard.js]: Fixed :)

File: src/controller/actions/topicDashboard.js (hunk 3 above)
Comment (jileon) [reply src/controller/actions/topicDashboard.js]: Fixed :)

**Pull request merged**